### PR TITLE
Allow `OCSPRequester` to fail `OCSPVerifierPolicy`

### PR DESCRIPTION
**Pull request by dnadoba**
### Motivation
A concrete `OCSPRequester` might want to decided that a request failure is fatal and therefore fail the OCSP verification.

### Modifications
Instead of throwing let the query method return a more descriptor return value that can disambiguate between terminal and non-terminal errors.

### Results
`OCSPRequester` can now result in a verification failure

### CODE REVIEW

File: Sources/X509/OCSP/OCSPPolicy.swift [hunk 1]
```diff
@@ -31,8 +31,51 @@ public protocol OCSPRequester: Sendable {
     /// - Parameters:
     ///   - request: DER-encoded request bytes
     ///   - uri: uri of the OCSP responder
-    /// - Returns: DER-encoded response bytes
-    func query(request: [UInt8], uri: String) async throws -> [UInt8]
+    /// - Returns: DER-encoded response bytes if they request was successful or a terminal or non-terminal error.
+    func query(request: [UInt8], uri: String) async -> OCSPRequesterQueryResult
+}
+
+public struct OCSPRequesterQueryResult: Sendable {
```
Comment (FranzBusch) [inline Sources/X509/OCSP/OCSPPolicy.swift:38]: `Hashable`?

File: Sources/X509/OCSP/OCSPPolicy.swift [hunk 2]
```diff
@@ -31,8 +31,51 @@ public protocol OCSPRequester: Sendable {
...
+}
+
+extension OCSPRequesterQueryResult {
+    /// The OCSP query is considered successful and has returned the given DER-encoded response bytes.
+    /// - Parameter bytes: DER-encoded response bytes
+    @inlinable
+    public static func response(_ bytes: [UInt8]) -> Self {
+        .init(.success(bytes))
+    }
+
```
Comment (FranzBusch) [inline Sources/X509/OCSP/OCSPPolicy.swift:61]:
```suggestion
```

File: Sources/X509/OCSP/OCSPPolicy.swift [hunk 3]
```diff
@@ -31,8 +31,51 @@ public protocol OCSPRequester: Sendable {
...
+    init(_ storage: Storage) {
+        self.storage = storage
+    }
+}
+
+extension OCSPRequesterQueryResult {
+    /// The OCSP query is considered successful and has returned the given DER-encoded response bytes.
+    /// - Parameter bytes: DER-encoded response bytes
+    @inlinable
+    public static func response(_ bytes: [UInt8]) -> Self {
```
Comment (FranzBusch) [inline Sources/X509/OCSP/OCSPPolicy.swift:58]: Hm, how do you actually get the bytes out of the result with the new API? Similar question on the other APIs.

Comment (dnadoba) [reply Sources/X509/OCSP/OCSPPolicy.swift:58]: Its just meant to be consumed internally by `OCSPVerificationPolicy`

**Comment by dnadoba** [comment]
API breakage expected:
```
15:18:17 1 breaking change detected in X509:
15:18:17   💔 API breakage: func OCSPRequester.query(request:uri:) has return type change from [Swift.UInt8] to X509.OCSPRequesterQueryResult
```

File: Sources/X509/OCSP/OCSPPolicy.swift (hunk 3 above)
Comment (FranzBusch) [reply Sources/X509/OCSP/OCSPPolicy.swift:58]: Ah so nobody is going to call `query` but rather we are calling it for them. So they only need to be able to construct the return type but not handle it.

Comment (dnadoba) [reply Sources/X509/OCSP/OCSPPolicy.swift:58]: exactly!

**Review by FranzBusch — APPROVED** [review]

File: Sources/X509/OCSP/OCSPPolicy.swift [hunk 4]
```diff
@@ -31,8 +31,50 @@ public protocol OCSPRequester: Sendable {
...
+        .init(.success(bytes))
+    }
+    
+    /// The OCSP query is considered unsuccessful but will **not** fail verification, neither in ``OCSPFailureMode/soft`` nor in ``OCSPFailureMode/hard`` failure mode.
+    /// The certificate is then considered to meet the ``OCSPVerifierPolicy``.
+    /// - Parameter reason: the reason why the OCSP query failed which may be used for diagnostics
+    @inlinable
+    public static func nonTerminalError(_ reason: Error) -> Self {
```
Comment (Lukasa) [inline Sources/X509/OCSP/OCSPPolicy.swift:66]: The doc string here should probably have a `warning` directive that calls out that if you use this the OCSP validation will be assumed to have _succeeded_.

**Review by Lukasa — CHANGES_REQUESTED** [review]

File: Sources/X509/OCSP/OCSPPolicy.swift (hunk 4 above)
Comment (dnadoba) [reply Sources/X509/OCSP/OCSPPolicy.swift:67]: Added in https://github.com/apple/swift-certificates/pull/108/commits/8c659ed5947b6de8886ee8bbb31b514aac13ea29

**Review by Lukasa — APPROVED** [review]

**Pull request merged**